tactics/class_tactics.ml

```diff
@@ -992,7 +991,8 @@ module Search = struct
                          (str "leaking evar " ++ int (Evar.repr ev) ++
                             spc () ++ pr_ev evm' ev);
                      acc && okev) evm' true);
-           let fgoals = Evd.shelve_on_future_goals shelved fgoals in
+           let evm = Evd.shelve_on_future_goals shelved evm in
+           let fgoals = Evd.save_future_goals evm in
```

Comment:
@mattam82 I'm not sure what this code is doing. I guess we could come up with a better API than `save`/`restore`. Can you explain a bit what action is needed here on future goals?